book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+
+Using Python, JavaScript, C++ or Rust, one could define many different a data
+structure that would correspond to our idea of "relation". Most of them would be
+an object/struct that has column names and types and then a generic array of
+arrays for rows.
+
+PRQL's type system should also be able to express relation as composed from
+primitive types, but have only one idiomatic way of doing so.
+
+In practice this means that builtin types include only primitives (int, text,
+bool, float), struct (for product), enum (for sum) and list (for repeating).
+
+An SQL row would translate to struct, and a relation would translate to a list
+of structs.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Physical layout
+
+_Logical type_ is user-facing the notion of a type that is the building block of
+the type system.
+
+_Physical layout_ is the underlying memory layout of the data represented by a
+variable.
+
+In many programming languages, physical layout of a logical type is dependent on
+the target platform. Similarly, physical layout of a PRQL logical type is
+dependent on representation of that type in target STS.
+
+```
+PTS logical type  --->  STS logical type  ---> STS physical layout
+```
+
+Note that STS types do not have a single physical layout. Postgres has a logical
+type `anyelement`, which is a super type of any data type. It can be used a
+function parameter type, but does not have a single physical layout so it cannot
+be used a in column declarations.
+
+For now, PRQL does not define a physical layout of any type. It is not needed
+since PRQL is not used for DDL (see section "Built-in primitives") or does not
+support raw access to underlying memory.
+
+As a consequence, results of a PRQL query cannot be robustly compared across
+DBMSs, since the physical layout of the result will vary.
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
+type float
+type bool
+type text
+type char
+type null
+
+# users-defined types
+type my_int = int
+
+# sum type (union or enum)
+type number = int | float
```

Comment:
It's a naive question, but I think it's hard to understand if pipe and OR use the same symbol `|`.
Couldn't we use some other symbol here?
(I don't know of any other symbol that would be better suited for OR except `||`, so it might be better to leave it as is.)

Reply:
Yes, I agree - using same syntax as pipe it not ideal.

I think having `||` is better.

Reply:
Wait, I just realized that we already use `or` for logical disjunction. Can we use it here?

Let's say we have two sets:
```
let a = {1, 2, 3}
let b = bool

# does this make sense?
let union_of_a_and_b = a or b

# or is the better?
let union_of_a_and_b = a + b
```

P.S. Python uses `|` (binary or) for unioning sets.

Reply:
I agree that it is better to unify either "`&&` and `||`" or "`and` and `or`".
I prefer the former, but given that it is easier to distinguish from `|`, maybe the latter is better.